clients/src/main/java/org/apache/kafka/common/internals/KafkaFutureImpl.java

```diff
@@ -160,7 +160,7 @@ private void maybeThrowCancellationException(Throwable cause) {
      * Waits if necessary for this future to complete, and then returns its result.
      */
     @Override
-    public T get() throws InterruptedException, ExecutionException {
+    public abstract T get(long timeout, TimeUnit unit) throws InterruptedException, ExecutionException, TimeoutException {
```

Comment:
Apologies if I misread, why not use the method `get` with timeout line 177?

Reply:
That makes sense! I think this change might be in here by accident -- I thought I had done that. Thanks for pointing that out. I'll remove it.

Reply:
Thanks for the follow-up. 

Trying to understand your use case. 

The `MetadataRequest` used to describe topics timed out. All retries were exhausted (4 lines) and a `TimeoutException`  (FQN `org.apache.kafka.common.errors.TimeoutException`) was propagated to the future (that is the future was completed exceptionally), and then propagated to the caller resulting in the behaviour observed.

```
2022-07-29 13:39:37.854  INFO 25843 --- [348aefeff-admin] org.apache.kafka.clients.NetworkClient   : [AdminClient clientId=L.DII.A-b1355e4a-b909-4da1-a832-dd3348aefeff-admin] Disconnecting from node 3 due to request timeout.
2022-07-29 13:39:37.854  INFO 25843 --- [348aefeff-admin] org.apache.kafka.clients.NetworkClient   : [AdminClient clientId=L.DII.A-b1355e4a-b909-4da1-a832-dd3348aefeff-admin] Cancelled in-flight METADATA request with correlation id 985 due to node 3 being disconnected (elapsed time since creation: 60023ms, elapsed time since send: 60023ms, request timeout: 30000ms)
2022-07-29 13:39:37.867 ERROR 25843 --- [-StreamThread-1] o.a.k.s.p.i.InternalTopicManager         : stream-thread [main] Unexpected error during topic description for L.DII.A-COGROUPKSTREAM-AGGREGATE-STATE-STORE-0000000003-changelog.
Error message was: org.apache.kafka.common.errors.TimeoutException: Call(callName=describeTopics, deadlineMs=1659101977830, tries=1, nextAllowedTryMs=1659101977955) timed out at 1659101977855 after 1 attempt(s)
2022-07-29 13:39:37.869  INFO 25843 --- [-StreamThread-1] o.a.k.s.p.internals.StreamThread         : stream-thread [L.DII.A-b1355e4a-b909-4da1-a832-dd3348aefeff-StreamThread-1] State transition from RUNNING to PENDING_SHUTDOWN
```

Therefore, it seems that:

- The failure of the "describe-topics" invocation was already correctly propagated via the future.
- This failure results of the time out of the underlying Metadata request **and** exhaustion of retries.

If the broker(s) were only temporarily unavailable, it seems increasing the number of retries may have helped?

Reply:
I'm not sure what you mean here -- for more on the problem, read https://issues.apache.org/jira/browse/KAFKA-14128